chanfunding: fee estimation based on change output in `CoinSelect`

bjarnemagnussen opened the pull request:
This PR adds a dust-limit to `CoinSelect` and lets the fee estimation account for if a change output is used or not. Any change output below the dust limit goes towards the fee instead.

It would also be useful for #4029, allowing better interoperability between the two coin-select algorithms `CoinSelect` and `CoinSelectSubtractFees`.

halseth commented:
Could you explain how this is different than just not adding the change output if the `changAmt` is below the limit?

I'm trying to understand what it actually changes..

halseth reviewed (commented):
Could you explain how this is different than just not adding the change output if the changAmt is below the limit?

I'm trying to understand what it actually changes..

bjarnemagnussen commented:
Sure, I will try to better explain the idea behind the changes in this PR!

I am not sure I misunderstand your question/doubts though, so let me know if there is something else. 

I was inspired a lot by the `CoinSelectSubtractFees` function to account for the fee calculation. You could argue some of it is duplicated code. But since there are a couple of todos left in both functions that may lead to a refactor, I think this more simple way allows for a better refactor later in the future if necessary.

I may have explained it a bit poorly regarding what happens in case the change output would be below the dust limit. So if the change output is below the dust limit respecting the calculated fee with a change output, then the change will not be used (basically set to 0) and the fee _without_ a change output will be used instead. Implicitly the change output then goes towards the fee.

halseth commented:
Aha, I see! Thanks for the explanation :) 

Would it be possible to extract the code such that it could be shared between the two methods instead of duplicating it?

bjarnemagnussen commented:
I have extracted the fee estimation algorithm from the two methods and refactored it into its own helper function with corresponding tests. Although the two methods feel quite similar, in fact the one subtracts the fee from the output and the other from the change, it is still difficult for me to see if there is a more clever way to unify those methods?

BTW, I kept the type of the change output hardcoded as a P2WKH and did not allow it to be changed via e.g. a functional option, since I am not sure how the TODO left in the code may handle it in the future.

halseth commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -52,11 +52,68 @@ func selectInputs(amt btcutil.Amount, coins []Coin) (btcutil.Amount, []Coin, err
+// estimates, one calculated using a change output and one without. The weight
+// added to the estimator from a change output is for a P2WKH output.
+func calculateFees(utxos []Coin, feeRate chainfee.SatPerKWeight) (
+	btcutil.Amount, btcutil.Amount, error,
```
weird line wrapping

bjarnemagnussen replied:
I wasn't sure if that was a formatting rule or not (as per https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#48-additional-style-constraints-on-top-of-gofmt). I will change it to not have the wrapping placed there.

halseth commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -68,53 +125,47 @@ func CoinSelect(feeRate chainfee.SatPerKWeight, amt btcutil.Amount,
+			newChange := overShootAmt - requiredFeeWithChange
+			if newChange > dustLimit {
+				changeAmt = newChange
+			}
```
I would add a default clause to handle the case `requiredFeeNoChange < overShootAmt < requiredFeeWithChange` for readability.

```go
// If sufficient funds were selected to cover the fee required to include a change
// output, the remainder will be our change amount.
case overShootAmt > requiredFeeWithChange:
    changeAmt = overShootAmt - requiredFeeWithChange

// Otherwise we have selected enough to pay for a tx without a change output.
default:
      changeAmt = 0
}


if changeAmt < dustLimit {
    changeAmt = 0
}

bjarnemagnussen replied:
Great!

halseth commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -134,62 +185,31 @@ func CoinSelectSubtractFees(feeRate chainfee.SatPerKWeight, amt,
-
 	// For a transaction with a change output, everything we don't spend
 	// will go to change.
+	newOutput := amt - requiredFeeWithChange
```
why move the line?

bjarnemagnussen replied:
Just for consistency, because above and below "output" comes before "change", so scanning the code is a little bit easier.

halseth commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -68,53 +125,47 @@ func CoinSelect(feeRate chainfee.SatPerKWeight, amt btcutil.Amount,
+		// If sufficient funds were selected to cover for a change
+		// output above the dust limit, we'll add the change output.
+		// Otherwise we'll go with the no change transaction.
+		case overShootAmt > requiredFeeWithChange:
```
could you do the extraction of common code into `calculateFees` and the functional change here in separate commits? Makes it easier to see how the changes impacts the tests.

bjarnemagnussen commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -52,11 +52,68 @@ func selectInputs(amt btcutil.Amount, coins []Coin) (btcutil.Amount, []Coin, err
+
+	// Estimate the fee required for a transaction with a change output.
+	// Assume that change output is a P2WKH output.
+	// TODO: Handle wallets that generate non-witness change addresses.
```
Ok!

bjarnemagnussen commented:
I have split up the commits into two separates: 
* one that refactors the code, and 
* another that aligns the fee calculation.  

I have also incorporated the requested changes for easier readability (nice!).

bjarnemagnussen commented:
Rebased

halseth commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -123,28 +123,55 @@ func CoinSelect(feeRate chainfee.SatPerKWeight, amt btcutil.Amount,
 		// round of coin selection.
-		if overShootAmt < requiredFee {
-			amtNeeded = amt + requiredFee
+		case overShootAmt < requiredFeeNoChange:
```
update comment for the case to refleclt this is now without change output.

bjarnemagnussen replied:
Comment changed to make it more clear!

halseth commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -123,28 +123,55 @@ func CoinSelect(feeRate chainfee.SatPerKWeight, amt btcutil.Amount,
+		case overShootAmt > requiredFeeWithChange:
+			changeAmt = overShootAmt - requiredFeeWithChange
+
+		// Otherwise we have selected enough to pay for a tx without a
```
👍

halseth reviewed (approved):
Great work on this one, I say we'll merge it for the next release 👍

halseth commented:
(need a rebase to trigger a new Travis build btw)

carlaKC commented on lnwallet/chanfunding/coin_select_test.go:
```diff
@@ -42,6 +50,107 @@ func fundingFee(feeRate chainfee.SatPerKWeight, numInput int, // nolint:unparam
+				t.Fatalf(err.Error())
+			}
+
+			if test.expectErr && err == nil {
```
could add an `expectedError` field to the test and `require.Equal`, you'd only need to define an `errUnsupportedInput` error since that's the only error we're testing for. Simplifies this a bit.

carlaKC commented on lnwallet/chanfunding/coin_select_test.go:
```diff
@@ -42,6 +50,107 @@ func fundingFee(feeRate chainfee.SatPerKWeight, numInput int, // nolint:unparam
+
+			// If we got an expected error, there is nothing more
+			// to test.
+			if test.expectErr {
+				return
+			}
+
+			// Check the calculated fee without change output.
+			if feeNoChange != test.expectedFeeNoChange {
+				t.Fatalf("expected feeNoChange %d, got %d",
+					test.expectedFeeNoChange, feeNoChange)
+			}
+
+			// Check the calculated fee with change output.
+			if feeWithChange != test.expectedFeeWithChange {
+				t.Fatalf("expected feeWithChange %d, got %d",
+					test.expectedFeeWithChange,
+					feeWithChange)
```
could shorten this all to just `require.Equal(t, test.expectedFee..., fee...)`, the error cases have 0 values anyway so they'd also pass.

carlaKC commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -52,6 +52,61 @@ func selectInputs(amt btcutil.Amount, coins []Coin) (btcutil.Amount, []Coin, err
+// calculateFees returns for the specified utxos and fee rate two fee
+// estimates, one calculated using a change output and one without. The weight
+// added to the estimator from a change output is for a P2WKH output.
+func calculateFees(utxos []Coin, feeRate chainfee.SatPerKWeight) (btcutil.Amount,
```
nice refactor!

carlaKC commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -123,28 +123,56 @@ func CoinSelect(feeRate chainfee.SatPerKWeight, amt btcutil.Amount,
+		// Obtain fee estimates both with and without using a change
+		// output.
+		requiredFeeNoChange, requiredFeeWithChange, err := calculateFees(
+			selectedUtxos, feeRate)
```
nit: `)` on newline

carlaKC commented on lnwallet/chanfunding/wallet_assembler.go:
```diff
@@ -289,9 +289,10 @@ func (w *WalletAssembler) ProvisionChannel(r *Request) (Intent, error) {
 			localContributionAmt = r.LocalAmt
 			selectedCoins, changeAmt, err = CoinSelect(
-				r.FeeRate, r.LocalAmt, coins,
+				r.FeeRate, r.LocalAmt, dustLimit, coins,
```
We're still checking `changeAmt > dustLimit` below this, could remove that check or add a sanity check before we add the change output? Since we're now checking dust in `CoinSelect`.

carlaKC reviewed (commented):
Very nice change, thanks @bjarnemagnussen!

Just wondering whether we want to remove the existing dust limit check now that we've moved it into coinselect? Otherwise just some nits.

bjarnemagnussen reviewed (commented):
Thanks for the review @carlaKC ! I have forced pushed the changes. See also my comments as replies!

halseth commented on lnwallet/chanfunding/coin_select.go:
```diff
@@ -105,8 +151,6 @@ func CoinSelect(feeRate chainfee.SatPerKWeight, amt btcutil.Amount,
 		// coin amount by the estimate required fee, performing another
 		// round of coin selection.
-		totalWeight := int64(weightEstimate.Weight())
-		requiredFee := feeRate.FeeForWeight(totalWeight)
```
seems like this won't compile?

bjarnemagnussen replied:
It compiles now!

halseth commented on lnwallet/chanfunding/wallet_assembler.go:
```diff
@@ -299,11 +299,16 @@ func (w *WalletAssembler) ProvisionChannel(r *Request) (Intent, error) {
 
+		// Sanity check: The addition of the outputs should not lead to the
+		// creation of dust.
+		if changeAmt <= w.cfg.DustLimit {
```
If this is a sanity check, should return an error instead?

bjarnemagnussen replied:
Ah yes, I have changed it to return an error instead.

carlaKC reviewed (commented):
Looking really good, just needs to be updated so that each commit builds! See `commit check` action that's failing.

bjarnemagnussen commented on lnwallet/chanfunding/coin_select_test.go:
```diff
@@ -42,6 +51,89 @@ func fundingFee(feeRate chainfee.SatPerKWeight, numInput int, // nolint:unparam
+			// Note: The error-case will have zero values returned
+			// for fees and therefore anyway pass the following
+			// requirements.
+
```
:+1:

bjarnemagnussen commented:
I have made the changes discussed above and now each commit compiles correctly!

halseth reviewed (approved):
Great works, I think this is ready! LGTM ✅

carlaKC reviewed (approved):
Great work on this one!
